src/sentry/monitors/rate_limit.py

```diff
@@ -0,0 +1,99 @@
+# > unmarshal the event payload to rate-limit on the monitor slug. The project_id
+# > is available in the payload headers.
+#
+# To allow the monitors system to take-advantage of Relay's quota system to be
+# used for rate-limiting, we can rate-limit per project by providing relay with
+# a quota for the DataCategory.MONITOR that is calculated based on how many
+# monitors the project has.
+#
+# There are some caveats to this:
+#
+# - For upserted monitors we need to give some amount of minimum quota, to
+#   allow monitors to be created even when there are 0 monitors within a
+#   project (since otherwise it would be completely rate-limited)
+#
+# - It is possible that one monitor may saturate the quota, thus rate-limiting
+#   all other monitors within that project. This is an accepted side-effect of
+#   doing rate-limiting this way.
+#
+# - We give each monitor a 'grace' number of allowed check-ins. Ideally each
+#   monitor should only be checking in once per minute, so the quota could just
+#   be the number of monitors in the project. However,
+
+
+# Monitor check-in limits are per minute. This maps to the smallest check-in
+# window that we support.
+QUOTA_WINDOW = 60
+
+# Determines how many check-ins per-minute will be allowed per monitor. This is
+# used when computing the QuotaConfig for the DataCategory.MONITOR (check-ins)
+#
+# These are the 'grace' check-ins as described above.
+ALLOWED_CHECK_INS_PER_MONITOR = 5
+
+# The minimum rate-limit per project for the DataCategory.MONITOR. This value
+# should be high enough that it allows for a large number of monitors to be
+# upserted without hitting the project rate-limit.
+ALLOWED_MINIMUM = 50
+
+
+def get_project_monitor_quota(
+    project: Project,
+    cache_bust=False,
+) -> Tuple[Optional[int], Optional[int]]:
+    """
+    Determines the rate-limit for monitor check-ins across a particular
+    project.
+
+    :return: A (window, limit) tuple. (None, None) indicates no rate-limit
+    """
+    if not features.has(
+        "organizations:monitors-quota-rate-limit", organization=project.organization
+    ):
+        return (None, None)
+
+    limit = None
+    cache_key = f"project:{project.id}:monitor-env-count"
+
+    # Cache rate-limit computation. This function will be called often by the
+    # Quotas system.
+    if not cache_bust:
+        limit = cache.get(cache_key)
+
+    if limit is None:
+        monitor_count = MonitorEnvironment.objects.filter(monitor__project_id=project.id).count()
+        limit = monitor_count * ALLOWED_CHECK_INS_PER_MONITOR
+        cache.set(cache_key, limit, 600)
+
+    return (ALLOWED_MINIMUM + limit, QUOTA_WINDOW)
+
+
+def update_monitor_quota(monitor_env: MonitorEnvironment):
+    """
+    When new monitor environments are created we recompute the per-project
+    monitor check-in rate limit QuotaConfig in relay.
+    """
+    project = Project.objects.get_from_cache(id=monitor_env.monitor.project_id)
```

Comment:
If you're worried about extra queries, might be worth doing https://github.com/getsentry/sentry/blob/fe7320639f7e496b11245f9a3963194417293015/src/sentry/event_manager.py#L429-L435

Reply:
Handled here

https://github.com/getsentry/sentry/pull/61310/files#diff-d060826a639e3ced30cbf38a97452f34b9a179c6aa700faccbc10a7277774e8eR68